### ISSUE TITLE
`jaxlibWithCuda` on 23.11 does not use CUDA

### ISSUE THREAD

**Issue by breakds** [NixOS/nixpkgs#282184]
### Describe the bug

When creating a dev shell with `jaxlibWithCuda` + `jax`, the resulting environment cannot use CUDA with `jax`.

### Steps To Reproduce
1. Create a `flake.nix`

   ```nix
   {
     description = "Provide extra Nix packages for Machine Learning and Data Science";
     inputs = {
       nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";

       utils.url = "github:numtide/flake-utils";
     };
     outputs = { self, nixpkgs, ... }@inputs: inputs.utils.lib.eachSystem [
       "x86_64-linux"
     ] (system:
       let pkgs = import nixpkgs {
             inherit system;
             config = {
               allowUnfree = true;
               cudaSupport = true;
               cudaCapabilities = [ "7.5" "8.6" ];
               cudaForwardCompat = false;
             };
           };
       in rec {
         devShells.default = pkgs.mkShell {
           name = "jax-dev";
           packages = [
             (pkgs.python3.withPackages (py-pkgs: with py-pkgs; [
               jax
               jaxlibWithCuda
             ]))
           ];
         };
       });
   }
   ```
   Note that I am using the `d2003f2223cbb8cd95134e4a0541beea215c1073` commit from nixos-23.11, which is the latest at this moment. However, I have tried the commit from a month ago and the problem was there as well.

2. Now, run `nix develop` with the above flake.
3. Run `python` and 

   ```console
   Python 3.11.6 (main, Oct  2 2023, 13:45:54) [GCC 12.3.0] on linux
   Type "help", "copyright", "credits" or "license" for more information.
   >>> import jax
   >>> jax.devices()
   An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
   [CpuDevice(id=0)]
   ```
   Note that it reports `NVIDIA GPU` is found, but `jaxlib` does not seem to have CUDA enabled. 

The above `flake.nix` can be found [here](https://github.com/nixvital/ml-pkgs/blob/debug/jaxWithCuda/flake.nix).

### Expected behavior
It is expected that `jax` and `jaxlib` run with CUDA and GPU appears in the device list.

### Additional context

I am running this on my NixOS machine with a Nvidia GPU, where nvidia-driver is properly installed. I am also an intensive `pytorch` user and it works perfectly.

Also I vaguely remember such combination works correctly a few months ago.

### Notify maintainers

I am not particularly sure about the maintainers so I might be wrong. Based on `jax`'s commit history and meta I am goingt to @SomeoneSerge and @samuela 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.79, NixOS, 23.11 (Tapir), 23.11.20240105.c1be43e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by breakds** [comment]
Let me know if I can provide more useful information. Any suggestion on how I can continue debugging this would also be very helpful.

Thanks in advance!

**Comment by samuela** [comment]
hmm interesting... ooc what gpu are you trying to use? have you confirmed that the cuda capabilities specified in your flake are compatible with your gpu?

**Comment by SomeoneSerge** [comment]
Also please gist the outputs of `LD_DEBUG=libs python ...` demonstrating which implementation of jaxlib ends up being loaded, and whether a search for `libcuda.so` is attempted

**Comment by breakds** [comment]
> hmm interesting... ooc what gpu are you trying to use? have you confirmed that the cuda capabilities specified in your flake are compatible with your gpu?

On this machine it is a 3060 Ti. I have been using `torchWithCuda` on this machine, with the samilar dev Shell, so I can confirm the cuda capabilities specified should work.

**Comment by breakds** [comment]
> Also please gist the outputs of `LD_DEBUG=libs python ...` demonstrating which implementation of jaxlib ends up being loaded, and whether a search for `libcuda.so` is attempted

Thanks. Let me put the stderr log here:

[jax.log](https://github.com/NixOS/nixpkgs/files/13996172/jax.log)

**Comment by breakds** [comment]
I think it is trying to load from `/nix/store/28r0fhcqqykrdfx5kqzwzqqjakn92n12-python3.11-jaxlib-0.4.20/`, but I am not sure whether it is the GPU-enabled one or not.

**Comment by samuela** [comment]
Also just to confirm: what's the output of 
```
echo $PYTHONPATH | awk 'BEGIN{RS=":"}{print}' | grep "jaxlib"
```
? Sometimes multiple jaxlib versions can get pulled in as transitive dependencies (though they shouldn't if everything is packaged correctly). Then python can end up loading jaxlib-without-cuda instead of jaxlib-with-cuda depending on the ordering in PYTHONPATH.

From the looks of your flake this shouldn't be the case, but I'm guessing that this is a reproduction distilled from a more complex flake/project?

**Comment by breakds** [comment]
> From the looks of your flake this shouldn't be the case, but I'm guessing that this is a reproduction distilled from a more complex flake/project?

Yes this is distilled from a real world flake.

> Sometimes multiple jaxlib versions can get pulled in as transitive dependencies

It seems that in my dev shell `PYTHONPATH` is not set. However, running

```python
import sys
print(sys.path)
```

produces:

```console
['', '/nix/store/p1zbyfrpj3hq50mxh5hmxl3kqpa2b1am-python3-3.11.6/lib/python311.zip', '/nix/store/p1zbyfrpj3hq50mxh5hmxl3kqpa2b1am-python3-3.11.6/lib/python3.11', '/nix/store/p1zbyfrpj3hq50mxh5hmxl3kqpa2b1am-python3-3.11.6/lib/python3.11/lib-dynload', '/nix/store/p1zbyfrpj3hq50mxh5hmxl3kqpa2b1am-python3-3.11.6/lib/python3.11/site-packages', '/nix/store/kwyj30nr4g40skc2m4nw0v2h1hh3miw5-python3-3.11.6-env/lib/python3.11/site-packages']
```

I looked into the `site-packages` of the `python3-3.11.6-3nv`:

```console
breakds@samaritan {~/projects/nixvital.org/ml-pkgs} $ lsd /nix/store/kwyj30nr4g40skc2m4nw0v2h1hh3miw5-python3-3.11.6-env/lib/python3.11/site-packages
󰌠 __pycache__                                 flatbuffers-23.5.26.dist-info   ml_dtypes                   opt_einsum-3.3.0.dist-info   six-1.16.0.dist-info
 _sysconfigdata__linux_x86_64-linux-gnu.py   jax                             ml_dtypes-0.3.1.dist-info   README.txt                   six.py
 absl                                        jax-0.4.20.dist-info            numpy                       scipy
 absl_py-1.4.0.dist-info                     jaxlib                          numpy-1.26.1.dist-info      scipy-1.11.3.dist-info
 flatbuffers                                 jaxlib-0.4.20.dist-info         opt_einsum                  sitecustomize.py
```

Apparently there is only one `jaxlib` being pulled in here.

**Comment by samuela** [comment]
Ok yeah that looks fine.. hmm

**Comment by breakds** [comment]
Btw, I tried two more things

1. Remove the capabilities constraint - still the same
2. Tried this on a 3090 machine, a 3080 machine and a 4090 machine. All of them produce exactly the same behavior.

**Comment by samuela** [comment]
Attempting to repro this but currently blocked on https://github.com/nix-community/nixGL/issues/157.

**Comment by breakds** [comment]
> Attempting to repro this but currently blocked on [nix-community/nixGL#157](https://github.com/nix-community/nixGL/issues/157).

Interesting. I had an attempt to run my nix-based ML project on an Ubuntu machine but wasn't able to get the nix dev shell to recognize the GPU a while ago. Have been using NixOS for development since then ...

**Comment by samuela** [comment]
Oh yeah it definitely works... mostly. This is the first real issue I've had with nixGL haha

**Comment by breakds** [comment]
Will give it a try! Meanwhile, I can try out any thoughts/ideas you might have to debug. Thanks!

**Comment by samuela** [comment]
Yeah my hope upon getting a repro is to bisect nixpkgs commits to find a known-good commit and then hopefully isolate the issue in a commit range. If you're feeling up to the task it would be quite useful!

@SomeoneSerge @ConnorBaker This would be a great use case for GPU-enabled tests and some CI infrastructure. tl;dr is that `jaxlibWithCuda` build has been green for a while but at some point it stopped actually using/recognizing GPUs.

**Comment by breakds** [comment]
> Yeah my hope upon getting a repro is to bisect nixpkgs commits to find a known-good commit and then hopefully isolate the issue in a commit range. If you're feeling up to the task it would be quite useful!

Got it. This looks like something I can try. It would probably take a while as I'd imagine CUDA will be built again and again.

**Comment by breakds** [comment]
Let me try to build from a few commits on `master` to start with.

**Comment by breakds** [comment]
Tried [f195a5e7fad77b5128ebfaba0a6112cd9ddca0d2](https://github.com/NixOS/nixpkgs/commit/f195a5e7fad77b5128ebfaba0a6112cd9ddca0d2) from about 3 months ago. Got different behavior:

```console
Python 3.10.12 (main, Jun  6 2023, 22:43:10) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import jax
>>> jax.devices()
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1705820658.921836 1555874 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[CpuDevice(id=0)]
>>>
I0000 00:00:1705820728.067464 1555874 tfrt_cpu_pjrt_client.cc:352] TfrtCpuClient destroyed.
```

**Comment by breakds** [comment]
Also confirmed that https://github.com/NixOS/nixpkgs/commit/58538146ed1288a96a0b90ab91528c6229819721 can reproduce the probem:

```console
Python 3.11.5 (main, Aug 24 2023, 12:23:19) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import jax
>>> jax.devices()
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
[CpuDevice(id=0)]
```

**Comment by breakds** [comment]
The above seems to suggest that, it is likely that in the past 3 months `jaxlibWithCuda` would not work with CUDA.

**Comment by breakds** [comment]
I found the a successful commit on `23.05`: https://github.com/NixOS/nixpkgs/commit/70bdadeb94ffc8806c0570eb5c2695ad29f0e421

Please note that this is not a commit from `master` and I believe it is using CUDA 11.7

```console
>>> import jax
>>> jax.devices()
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0)]

>>> import jax.lib
>>> backend = jax.lib.xla_bridge.get_backend()
>>> backend.platform
'gpu'
>>> backend.platform_version
'cuda 11070'
```

**Comment by samuela** [comment]
Nice! Well I guess we have a (rough) commit window now to bisect on.

Btw, I'm curious: does `python3Packages.jaxlib-bin` work any better for you on master? We don't have a CUDA alias for that one, so you'll need to set `config.cudaSupport = true`

**Comment by breakds** [comment]
> Nice! Well I guess we have a (rough) commit window now to bisect on.
> 
> Btw, I'm curious: does `python3Packages.jaxlib-bin` work any better for you on master? We don't have a CUDA alias for that one, so you'll need to set `config.cudaSupport = true`

Let me try that. My understanding is that 

1. `jaxlib-bin` is about downloading the pre-built binary (wheel) and patch it, while
2. `jaxlibWithCuda` is about spinning up `bazel` to build the wheel by ourselves

And for end users of `jaxlib`, they should behave pretty much the same. Is my understanding correct? Thanks!

**Comment by breakds** [comment]
> Btw, I'm curious: does `python3Packages.jaxlib-bin` work any better for you on master? We don't have a CUDA alias for that one, so you'll need to set `config.cudaSupport = true`

Tried `jaxlib-bin`. Seeing the following error:

```console
>>> import jaxlib.xla_client
2024-01-22 16:27:55.773987: I external/tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
```

**Comment by breakds** [comment]
Digged a little bit into this.

On the successful branch 23.05 with CUDA 11.7:

```console
>>> import jaxlib
>>> import jaxlib.xla_client
>>> jaxlib.xla_client.make_gpu_client(platform_name="cuda")
2024-01-22 17:11:18.541943: I external/org_tensorflow/tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:996] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-22 17:11:18.542124: I external/org_tensorflow/tensorflow/compiler/xla/service/service.cc:169] XLA service 0xabbb80 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2024-01-22 17:11:18.542155: I external/org_tensorflow/tensorflow/compiler/xla/service/service.cc:177]   StreamExecutor device (0): NVIDIA GeForce RTX 3060 Ti, Compute Capability 8.6
2024-01-22 17:11:18.542372: I external/org_tensorflow/tensorflow/compiler/xla/pjrt/gpu/se_gpu_pjrt_client.cc:198] Using BFC allocator.
2024-01-22 17:11:18.542407: I external/org_tensorflow/tensorflow/compiler/xla/pjrt/gpu/gpu_helpers.cc:105] XLA backend allocating 6270959616 bytes on device 0 for BFCAllocator.
<jaxlib.xla_extension.Client object at 0x7ffee56837b0>
```

while on the failed branch 23.11,

```console
>>> import jaxlib
>>> import jaxlib.xla_client
>>> jaxlib.xla_client.make_gpu_client(platform_name="cuda")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/1z91x9lrhrv195vbhl4962qclmnmnwhv-python3-3.11.6-env/lib/python3.11/site-packages/jaxlib/xla_client.py", line 88, in make_gpu_client
    config = _xla.GpuAllocatorConfig()
             ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'jaxlib.xla_extension' has no attribute 'GpuAllocatorConfig'
```

**Comment by samuela** [comment]
> And for end users of jaxlib, they should behave pretty much the same. Is my understanding correct? Thanks!

Yes, that's exactly right!

> This seems to suggest that the jaxlib we have in this case is not enabled for cuda support.

Nice find! Huh, very weird. We're building with all the GPU/CUDA flags turned on AFAIK. Also, it's surprising to me that jaxlib-bin isn't working on your system since upstream builds that and we make only minor modifications

**Comment by breakds** [comment]
> Nice find! Huh, very weird. We're building with all the GPU/CUDA flags turned on AFAIK.

I checked their [build script](https://github.com/google/jax/blob/main/build/build.py#L303) and found that there is a bazel flag called `cuda_plugin`. Don't know what it is (maybe related to [this](https://github.com/pytorch/xla/blob/master/docs/pjrt.md)?), but `jaxlib`'s `default.nix` does not have it. I tried to add it and `jaxlib` (more specifically, the `bazel-build` inside it) cannot build any more.

>  Also, it's surprising to me that jaxlib-bin isn't working on your system since upstream builds that and we make only minor modifications

Sorry if I did not make it clear. `jaxlib-bin` can be successfully built on my system. The problem is similar to `jaxlib`, as it cannot be used with CUDA. It just prompts a different error.

**Comment by breakds** [comment]
More on `jaxlib-bin`. Although it can successfully create the "client", `cudart_stub.cc` will report not being able to find cuda drivers:

```console
>> import jaxlib.xla_client
2024-01-23 13:51:38.735183: I external/tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
>>> jaxlib.xla_client.make_gpu_client(platform_name="cuda")
2024-01-23 13:54:38.390979: I external/xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must b
e at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-23 13:54:38.391106: I external/xla/xla/service/service.cc:168] XLA service 0xb9f6f0 initialized for platform CUDA (this does not guarantee that XLA will be used
). Devices:
2024-01-23 13:54:38.391129: I external/xla/xla/service/service.cc:176]   StreamExecutor device (0): NVIDIA GeForce RTX 3080, Compute Capability 8.6
2024-01-23 13:54:38.391319: I external/xla/xla/pjrt/gpu/se_gpu_pjrt_client.cc:738] Using BFC allocator.
2024-01-23 13:54:38.391342: I external/xla/xla/pjrt/gpu/gpu_helpers.cc:105] XLA backend allocating 7869579264 bytes on device 0 for BFCAllocator.
<jaxlib.xla_extension.Client object at 0x7ffee6a335b0>
```

This is `jaxlib-bin` on 23.11.

**Comment by breakds** [comment]
Tracing the code and I realized that the problem is that this line https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/platform/default/dso_loader.cc#L57, which in turn calls `dlopen` in https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/platform/default/load_library.cc#L54 failed for `libcudart.so.<version>`. 

Thanks to @SomeoneSerge 's advice, I added `cuda_cudart` here at https://github.com/NixOS/nixpkgs/blob/nixos-23.11/pkgs/development/python-modules/jaxlib/bin.nix#L184 in the `patchelf --set-rpath`. However, the issue persists for `jaxlib-bin`.

**Comment by samuela** [comment]
> Sorry if I did not make it clear. jaxlib-bin can be successfully built on my system. The problem is similar to jaxlib, as it cannot be used with CUDA. It just prompts a different error.

Yup, totally with you there. Now if jaxlib-bin didn't build successfully that'd be an easier fix... Debugging with jaxlib-bin can be convenient since it offers "known-good" binaries to work with.

> Based on [this](https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/third_party/gpus/cuda_configure.bzl#L731), I think the here should be 11.0. And it turns out that [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/pkgs/development/python-modules/jaxlib/bin.nix#L184) we already have patchelf for cudatoolkit.lib, which contains libcudart.so.11.0.

Hmm, ok this is where `LD_DEBUG=libs` output ought to point the way. But looking through the logs you posted earlier it, I'm not able to find the string `libcudart` anywhere... Would you mind posting the `LD_DEBUG=libs` logs for the `jaxlib-bin` version?

OOC how did you come to conclude that libcudart.so.<version> is the failing lookup? It's a perfectly sensible conclusion since just about every CUDA package ever has to load it, but just curious re your methodology.

**Comment by breakds** [comment]
> OOC how did you come to conclude that libcudart.so. is the failing lookup? It's a perfectly sensible conclusion since just about every CUDA package ever has to load it, but just curious re your methodology.

Ah sorry I should have provided more information here. So my journey starts with the error:

```
external/tsl/tsl/cuda/cudart_stub.cc:32] Could not find cuda drivers on your machine, GPU will not be used.
```

and this quickly leads to the code in tensorflow: https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/cuda/cudart_stub.cc#L32

This means that the call to `tsl::internal::DsoLoader::GetCudaRuntimeDsoHandle` failed. This particular function can be found at [here](https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/platform/default/dso_loader.cc#L91), which is essentially `GetDsoHandle("cudart", GetCudaRtVersion())`.

The function `GetDsoHandle` is [in the same file](https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/platform/default/dso_loader.cc#L52), which is basically a call to `tsl::internal::LoadDynamicLibrary(filename.c_str(), &dso_handle);`, which in turn is just a call to `dlopen` of the `filename` here.

Reading the above functions, we can see that the offending filename is in the format of `libcudart.so.<version>`. I could not figure out the `version` because it is actually defined [here](https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/platform/default/dso_loader.cc#L42), which is in turn injected from [bazel config](https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/third_party/gpus/cuda_configure.bzl#L731). Because I had the impression that I was using `11.8`, so I believe the version here is `11.0`.

This turned out to be wrong, see below.


> Hmm, ok this is where LD_DEBUG=libs output ought to point the way. But looking through the logs you posted earlier it, I'm not able to find the string libcudart anywhere... Would you mind posting the LD_DEBUG=libs logs for the jaxlib-bin version?

Thanks a lot for the suggestion! I tried the following

```consile
$ LD_DEBUG=libs python -c "import jaxlib.xla_extension" 2> ~/tmp/debug_libs_jaxlib.txt
```

Which produces log:

[debug_libs_jaxlib.txt](https://github.com/NixOS/nixpkgs/files/14034043/debug_libs_jaxlib.txt)

Turns out that it is actually looking for `libcudart`. However, it is looking for `libcudart.so.12`! Below are the selected part of the above log:

```
    999370:	find library=libcudart.so.12 [0]; searching
    999370:	 search path=/run/opengl-driver/lib:/run/opengl-driver-32/lib		(LD_LIBRARY_PATH)
    999370:	  trying file=/run/opengl-driver/lib/libcudart.so.12
    999370:	  trying file=/run/opengl-driver-32/lib/libcudart.so.12
    999370:	 search path=/run/opengl-driver/lib		(LD_LIBRARY_PATH)
    999370:	  trying file=/run/opengl-driver/lib/libcudart.so.12
    999370:	 search path=/nix/store/4zp05qkzh9s3c3k7pp9d36vhiz3x8h9l-gcc-13.2.0-lib/lib:/nix/store/3dimjffv8pn3sk48ngd4adz17agpfb4y-gcc-13.2.0-libgcc/lib:/nix/store/1fg4mqmsk3rfqa8psqc41qwlkk3r652z-cudatoolkit-11.8.0/lib:/nix/store/mdgi51z42vymqyck7m324hr876hw5ara-cudatoolkit-11.8.0-lib/lib:/nix/store/irsvvw6s40m6p5w491hha1nc5pji79i9-cudnn-8.9.7.29/lib		(RUNPATH from file /nix/store/asiphbpiy2gmidfm3xbwcikayhs66289-python3-3.11.7/bin/python)
    999370:	  trying file=/nix/store/4zp05qkzh9s3c3k7pp9d36vhiz3x8h9l-gcc-13.2.0-lib/lib/libcudart.so.12
    999370:	  trying file=/nix/store/3dimjffv8pn3sk48ngd4adz17agpfb4y-gcc-13.2.0-libgcc/lib/libcudart.so.12
    999370:	  trying file=/nix/store/1fg4mqmsk3rfqa8psqc41qwlkk3r652z-cudatoolkit-11.8.0/lib/libcudart.so.12
    999370:	  trying file=/nix/store/mdgi51z42vymqyck7m324hr876hw5ara-cudatoolkit-11.8.0-lib/lib/libcudart.so.12
    999370:	  trying file=/nix/store/irsvvw6s40m6p5w491hha1nc5pji79i9-cudnn-8.9.7.29/lib/libcudart.so.12
```

I then went back to check for the `jaxlib/bin.nix`: https://github.com/NixOS/nixpkgs/blob/nixos-23.11/pkgs/development/python-modules/jaxlib/bin.nix#L134, and apparently it is downloading the prebuilt binary that is tagged with `cuda12`. Note that `cudaPackagesGoogle` points to Cuda 11.8.

I might be wrong, but is prebuilt binary tagged with `cuda12` supposed to work with CUDA 11 as well? The `dlopen` logic that finds and loads the `.so` files definitely is not compatible with CUDA 11.

**Comment by SomeoneSerge** [comment]
> Note that cudaPackagesGoogle points to Cuda 11.8.

For as long as tf is 2.13: https://www.tensorflow.org/install/source#gpu. We should add the comment.

It's possible that jaxlib and tf actually come with different version constraints, I haven't checked.

> and apparently it is downloading the prebuilt binary that is tagged with cuda12. Note that

I guess one then has to pass it (`jaxlib-bin` specifically) `cudaPackages_12`.

> and this quickly leads to the code in tensorflow: https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/third_party/tsl/tsl/cuda/cudart_stub.cc#L32

Looks like jaxlib is trying to take over a part of the dynamic linker's job instead of just using DT_NEEDED and naturally this doesn't play out smoothly. They probably wanted to defer loading or something

**Comment by breakds** [comment]
> I guess one then has to pass it (jaxlib-bin specifically) cudaPackages_12.

I am trying adding an overlay like this:

```nix
final: prev: {
  cudaPackagesGoogle = prev.cudaPackages_12;
}
``

Seems that `jaxlib-bin` is now working. I am not sure whether this will have bad implications though.

> They probably wanted to defer loading or something

Yeah they might have some reason to defer it.

**Comment by breakds** [comment]
This is one problem though, if I have the following overlay:

```nix
final: prev: {
  cudaPackagesGoogle = prev.cudaPackages_12;
  
  pythonPackagesExtensions = prev.pythonPackagesExtensions ++ [
    (python-final: python-prev: {
      jax = python-prev.jax.override {
        jaxlib = python-final.jaxlib-bin;
      };
    })
  ];
}
```

`jaxlib` is ok, but `jax` will fail to build because of test failures.

**Comment by breakds** [comment]
At this moment, with the help of @samuela and @SomeoneSerge (Thanks!), I think I might have a solution to this.

## What is clear?

TLTR: We can now use `jaxlib-bin` (version 0.4.23) on `master`, if you apply an overlay to point `cudaPackagesGoogle` to `cudaPackages_12_2`.

**NOTE**: usually you will also need `jax` alongside `jaxlib-bin`. It works on `master`, but you will have to disable the unit tests and import test. See below in "what is still unclear".

Detailed explanation:

1. `jaxlib-bin` download a prebuilt binary, which defers loading of `libcudart.so.<version>` at runtime via `dlopen` dynamically.
2. The prebuilt binary is built against `CUDA 12.2`, meaning that `dlopen` will look for `libcudart.so.12`.
3. Also, there is another check to make sure that the CUDA version is exactly `12.2` (i.e. 12.1 and 12.0 will not work).
4. Currently, `jaxlib-bin` uses `cudaPacakgesGoogle`, which is designed for `tensorflow`. The current `tensorflow` needs `11.8` instead of `12.2`. However, if your environment does not require `tensorflow`, just point `cudaPackagesGoogle` to `12.2` and everything should work out. 

## What is still unclear?

1. Even with `jaxlib-bin` fixed above, I cannot build `jax` unless all the tests are disabled. It does not seem to affect the actual usage of `jax` because the test failure is specifically about not finding GPU-version of `jaxlib` during test. This should be fixable but I am not sure how at this moment.
2. Ideally we want to be able to use `jaxlib` instead of `jaxlib-bin` to have more flexibility. The current build of `jaxlibWithCuda` seems to produce a library that does not have CUDA support. This remains a mystery and needs further investigation.

**Comment by samuela** [comment]
I propose that we template the "cudaXX" strings in 
https://github.com/NixOS/nixpkgs/blob/8916302f6b5b3e4c17f3a0f53210d8683d1c7c98/pkgs/development/python-modules/jaxlib/bin.nix#L121-L141
based on the version of the inputted `cudaPackages`/`cudaPackagesGoogle` instead of hardcoding. That will prevent this kind of issue in the future. Unfortunately I'm of minimal use in terms of submitting PRs until https://github.com/nix-community/nixGL/issues/157 is resolved.

This reminds me that we should also migrate jaxlib-bin to use the redist packages at some point.

Fixing the source build will be its own adventure. I suspect we're just missing some kind of flag that changed in upstream's build process.

**Comment by breakds** [comment]
> based on the version of the inputted cudaPackages/cudaPackagesGoogle instead of hardcoding

Sounds good. Still, for the upstream releases, cuda11 = cuda 11.8, cuda12 = cuda 12.2. The minor version is still important here and should be checked.

> Fixing the source build will be its own adventure. I suspect we're just missing some kind of flag that changed in upstream's build process.

Agree. It probably requires setting a development environment and build incrementally with `bazel` to investigate. Blindly update flag and build the whole package again from scratch is quite inefficient ...

**Comment by breakds** [comment]
> It's worth noting that upstream releases builds for both cuda11 and cuda12
Sorry but I was not able to find the build for `cuda11`. I saw the ones under `cuda11_plugins`, but not under `cuda11`. Are they the same?

**Comment by samuela** [comment]
> Sorry but I was not able to find the build for cuda11. I saw the ones under cuda11_plugins, but not under cuda11. Are they the same?

Search https://storage.googleapis.com/jax-releases for "cuda11/jaxlib-0.4.23" and you'll find them. This link
https://github.com/NixOS/nixpkgs/blob/559ddda570f61c4b66fb53162db7ce781c818f9f/pkgs/development/python-modules/jaxlib/bin.nix#L121
seems to be outdated, or at least not actively maintained by upstream. We should prob update the comment to point to https://storage.googleapis.com/jax-releases instead.

**Comment by SomeoneSerge** [comment]
> Unfortunately I'm of minimal use in terms of submitting PRs until https://github.com/nix-community/nixGL/issues/157 is resolved.

@samuela I saw you're using that with Ubuntu? Did you try nixglhost?

**Comment by samuela** [comment]
> @samuela I saw you're using that with Ubuntu? Did you try nixglhost?

oooh no i hadn't heard of nixglhost before... i'll give that a shot!

**Comment by breakds** [comment]
> Search https://storage.googleapis.com/jax-releases for "cuda11/jaxlib-0.4.23" and you'll find them

With that I was able to find the one for CUDA 11 (cudnn 8.6) and it worked out with this overlay: https://github.com/nixvital/ml-pkgs/blob/dev/cuda11/overlays/jax-family.nix

Need some advice on how to fix `jaxlib-bin` on `master`. Based on @samuela's suggestion, I am thinking about

1. Check the version of cuda, assert on 11.8 / 12.2
2. `cudnn` will be set to `8.6` if cuda is 11.8, or `8.9` if cuda is 12.2
3. Fetch the `.whl` file based on the above version

Does that make sense?

**Comment by samuela** [comment]
> Does that make sense?

Yup, exactly what I was thinking! Though I have a sneaking suspicion that CUDA versions outside of 11.8 / 12.2 might work. Sometimes they'll attempt to `dlopen` `libcudart.so.11.8` and then try `libcudart.so.11`, and so forth. So we should try seeing if other CUDA versions work too.

**Comment by breakds** [comment]
> > Does that make sense?
> 
> Yup, exactly what I was thinking! Though I have a sneaking suspicion that CUDA versions outside of 11.8 / 12.2 might work. Sometimes they'll attempt to `dlopen` `libcudart.so.11.8` and then try `libcudart.so.11`, and so forth. So we should try seeing if other CUDA versions work too.

Thanks! I tried 12.0 and it won't work for `jaxlib` prebuilt binary originally built against 12.2.

**Comment by samuela** [comment]
Ah ok maybe not then... damn

**Comment by SomeoneSerge** [comment]
> What is clear? ... What is still unclear?

I still don't understand (and haven't got around to verify myself) whether `jaxlib` is functional, and why do we keep focusing on `jaxlib-bin` instead?

**Comment by breakds** [comment]
> > What is clear? ... What is still unclear?
> 
> I still don't understand (and haven't got around to verify myself) whether `jaxlib` is functional, and why do we keep focusing on `jaxlib-bin` instead?

`jaxlib` is not functional, which is the original report of this issue:
- `jaxlib` can be built
- However, the built `jaxlib` (more specifically, `jaxlibWithCuda`), was not able to work with CUDA.
- It seems that the problem is `jaxlibWithCuda` is actually not "withCuda". There might be some flags to be set during bazel build, but I was not able to find the reason.

We then turn to `jaxlib-bin` as an alternative because we need a working `jaxlib` at this moment. Turns out that `jaxlib-bin` has a different issue, but is easier to fix.

**Comment by breakds** [comment]
> the JAX folks claim that this ought to be possible: https://jax.readthedocs.io/en/latest/installation.html#nvidia-gpu

I also read about that, but my experiment wasn't successful, for example, it will panic about JAX being built with 12020 but the current CUDA version is 12000

**Comment by SomeoneSerge** [comment]
Looking at `nix derivation show` for https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fp52nf36wdkhflbb61p85n051rnk3cmzm-bazel-build-jaxlib-0.4.23.drv?via-job=3ef9a959-e327-4013-94d3-33ba5fb600f7 I see no mention of `--config=cuda` (https://github.com/google/jax/blob/a6f26306b387ce082358e8e94fdcfd777773475f/build/build.py#L294-L295)

Is it intentional we add `--config=cuda` to `fetchAttrs.bazelFlags` but not to the following? https://github.com/NixOS/nixpkgs/blob/2047e2ce33c6c14921df01a3d71e9faee80e2017/pkgs/development/python-modules/jaxlib/default.nix#L215

**Comment by breakds** [comment]
Similar here, I am not that familiar with jax's bazel `BUILD` code. Would like to provide some information here.

`jaxlib-0.4.4` with CUDA 11.7 works, and at that time `jaxlib/default.nix` looks very similar:

https://github.com/NixOS/nixpkgs/blob/bf9e6fe9b852ea9185b9d495dc974491d4fbf355/pkgs/development/python-modules/jaxlib/default.nix#L198-L220

**Comment by breakds** [comment]
> > Does that make sense?
> 
> Yup, exactly what I was thinking! Though I have a sneaking suspicion that CUDA versions outside of 11.8 / 12.2 might work. Sometimes they'll attempt to `dlopen` `libcudart.so.11.8` and then try `libcudart.so.11`, and so forth. So we should try seeing if other CUDA versions work too.

It seems that the version of cudnn is not a requirement. Binary built with `cudnn` 8.6 can work with 8.9 at runtime.

**Comment by samuela** [comment]
> Is it intentional we add --config=cuda to fetchAttrs.bazelFlags but not to the following?

Well, there's definitely a bug in here somewhere. It is def possible that this is part of it. I'm guessing that upstream changed their build flags and we completed a version upgrade without realizing since we don't have a means to run GPU-enabled tests in CI/nixpkgs-review yet.

`--config=cuda` is included in `fetchAttrs.bazelFlags` as a means to reduce the number of possible `fetchAttrs` hashes that we have to record. The `--config=cuda` fetch is a superset of the CPU-only fetch, so we combine them for simplicity.

> why do we keep focusing on jaxlib-bin instead?

`jaxlib-bin` is really handy in these situations where the source build is either broken or not on a version that the user wants. The source build is really quite unwieldy unfortunately. (Similar but not quite as bad as the tensorflow situation.)

**Comment by breakds** [comment]
I am working on the fix for `jaxlib-bin`. Not sure what is the best way to test it though. I am thinking about:

1. Create the fix in a branch
2. Create a `flake.nix` that points to that branch
3. Test `jaxlib-bin` with CUDA 11 and 12 on 3.9 through 3.12

**Comment by breakds** [comment]
The PR is created at https://github.com/NixOS/nixpkgs/pull/285037

**Comment by breakds** [comment]
I am closing this as at least the `bin` version is fixed now. Thanks a lot for the help @samuela @SomeoneSerge!

**Comment by samuela** [comment]
Fixing the *-bin version is great, but let's keep this open until the source version is fixed as well

**Comment by rekado** [comment]
FWIW, the Guix version (there is only a source version) is here: https://github.com/guix-science/guix-science-nonfree/blob/master/guix-science-nonfree/packages/machine-learning.scm#L216
It works fine with CUDA.

The definition of the CUDA-less version that it references is here:
https://github.com/guix-science/guix-science/blob/master/guix-science/packages/python.scm#L548

Maybe there's an obvious difference somewhere in the arguments.

**Comment by samuela** [comment]
Ok I think I've found the culprit. Somewhere along the way we removed `--config=cuda` from `bazelFlags` and/or `.jax_configure.bazelrc`. Oops!

Working on a fix now... Currently blocked on https://github.com/google/jax/issues/19811#issuecomment-1944467666. AFAICT we do not package CUB and it is not provided in any of the cudaPackages that I've consulted thus far. If anyone has any ideas on the best way to proceed, do let me know!

**Comment by SomeoneSerge** [comment]
> AFAICT we do not package CUB and it is not provided in any of the cudaPackages that I've consulted thus far. If anyone has any ideas on the best way to proceed, do let me know!

Pretty sure it's part of `cccl`, and we also used to distribute it as part of nvidia-thrust which we deleted

**Comment by samuela** [comment]
Oop, found `cudaPackages.cuda_cccl`, but not finding that header file sadly:
```
ubuntu@bitbop:~/nixpkgs$ nix-build -A cudaPackagesGoogle.cuda_cccl
these 2 paths will be fetched (1.15 MiB download, 12.74 MiB unpacked):
  /nix/store/r9677wlkn8bq9idbr2f3gwqpnjzagp4h-cuda_cccl-11.8.89
  /nix/store/30cq1zsb0a5a2xkca1jkn7gliyjbswsg-cuda_cccl-11.8.89-dev
copying path '/nix/store/30cq1zsb0a5a2xkca1jkn7gliyjbswsg-cuda_cccl-11.8.89-dev' from 'https://cuda-maintainers.cachix.org'...
copying path '/nix/store/r9677wlkn8bq9idbr2f3gwqpnjzagp4h-cuda_cccl-11.8.89' from 'https://cuda-maintainers.cachix.org'...
/nix/store/r9677wlkn8bq9idbr2f3gwqpnjzagp4h-cuda_cccl-11.8.89
ubuntu@bitbop:~/nixpkgs$ sudo find /nix -type f -name "cublock_load.cuh"
ubuntu@bitbop:~/nixpkgs$ 
```

**Comment by SomeoneSerge** [comment]
>    20 | #include "cub/block/block_load.cuh"

Not "cublock_load.cuh"?

```
❯ nix build -f . --arg config '{ allowUnfree = true; cudaSupport = true; }' cudaPackages.cuda_cccl -o cccl -L
❯ fd block_load cccl/
cccl/include/cub/block/block_load.cuh
```

**Comment by SomeoneSerge** [comment]
> packaging CUB from scratch

https://github.com/NixOS/nixpkgs/issues/224292

**Comment by samuela** [comment]
Adding `cudaPackages.cuda_cccl.dev` seemed to solve the problem, despite the difference in filenames. Not sure exactly what kind of wizardry they're doing to adjust header names, but it seems happy for now.

**Comment by SomeoneSerge** [comment]
> despite the difference in filenames

Looking at your issue upstream, I think the names match?

**Comment by samuela** [comment]
> Looking at your issue upstream, I think the names match?

Oop, yeah you're right. I was confusing that with something else

**Comment by CMCDragonkai** [comment]
If you're using NixOS is there a platform wide config to always have the cuda libraries enabled?

**Comment by aryanjassal** [comment]
I'm trying to compile jax using the commit `cdd38b2`, which appears to have closed this issue. When I tried to run jax on CUDA using this commit, I would need to compile jaxlib from scratch. During the compilation, I get a hash mismatch error in a library that `jaxlib` is compiling.
```console
[user@system:~/jax]$ nix-shell
[...snip...]
INFO: Analyzed 2 targets (240 packages loaded, 20380 targets configured).
 checking cached actions
INFO: Found 2 targets...
INFO: Elapsed time: 115.749s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 0 total actions
buildPhase completed in 1 minutes 56 seconds
Running phase: installPhase
installPhase completed in 49 seconds
error: hash mismatch in fixed-output derivation '/nix/store/yjm7vri9hzasddanrjhzd4fi1a9406ip-bazel-build-jaxlib-0.4.24-deps.tar.gz.drv':
         specified: sha256-IEKoHjCOtKZKvU/DUUjbvXldORFJuyO1R3F6CZZDXxM=
            got:    sha256-DLx5NHoSP9cOCf6hBgLXqNDb/XhJ3OWxA8/HGiTBBUo=
error: 1 dependencies of derivation '/nix/store/dsd7ssdqbmi82fvj1fcf8pnzv95wf6c4-bazel-build-jaxlib-0.4.24.drv' failed to build
error: 1 dependencies of derivation '/nix/store/yivr8ykzvg43g33hlm87f400568vd9ly-python3.11-jaxlib-0.4.24.drv' failed to build
error: 1 dependencies of derivation '/nix/store/vhr8syx8h5s3j18bp4kw9ifr3wviqd8p-python3.11-jax-0.4.24.drv' failed to build
```

I even tried to use a later commit `d97b37` to get the following output:
```console
[user@system:~/jax]$ TF_CPP_MIN_LOG_LEVEL=0 python -c "from jax.lib import xla_bridge; print(xla_bridge.get_backend().platform)"
2024-06-17 06:08:48.914216: I external/tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2024-06-17 06:08:48.970135: I external/tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2024-06-17 06:08:48.970677: I external/tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2024-06-17 06:08:49.123745: I external/tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
CUDA backend failed to initialize: Unable to load CUDA. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
cpu
```

This is my shell.nix:
```nix
{ pkgs ? import ./pkgs.nix { inherit builtins; } }:

pkgs.mkShell {
  name = "jax-gpu";

  buildInputs = with pkgs; with pkgs.python3Packages; with pkgs.cudaPackages; [
    python3
    jax
    jaxlib-bin
    numpy
    cuda_cccl.dev
  ];
}
```

And this is my pkgs.nix (for the newer commit):
```nix
{ builtins }:

let
  revision = "d97b37430f8f0262f97f674eb357551b399c2003";
  url = "https://github.com/nixos/nixpkgs/archive/${revision}.tar.gz";

  nixpkgs = builtins.fetchTarball {
    url = url;
    sha256 = "0za8bwgvk83r18z1b7ll1hlh45ijimp388ll5r8z44yglfcp28gx";
  };
in
  import nixpkgs {
    config = {
      allowUnfree = true;
      cudaSupport = true;
    };
  }
```

**Comment by SomeoneSerge** [comment]
> If you're using NixOS is there a platform wide config to always have the cuda libraries enabled? @CMCDragonkai 

This seems like a support question, you're more likely to find help on NixOS Discourse or Matrix. But the answer is `nixpkgs.config.cudaSupport = true` and this will trigger a lot of rebuilds

> I get a hash mismatch error in a library that jaxlib is compiling. @aryanjassal 

Interesting, this might be a reproducibility issue with the way we fetch bazel dependencies, in which case this deserves a separate issue
